multiversx_sdk_core/transaction_factories/smart_contract_transactions_factory.py

```diff
@@ -57,24 +59,83 @@ def create_transaction_for_deploy(self,
 
         return transaction
 
-    def create_transaction_intent_for_execute(self,
-                                              sender: IAddress,
-                                              contract_address: IAddress,
-                                              function: str,
-                                              gas_limit: int,
-                                              arguments: List[Any] = []) -> Transaction:
-        parts = [function] + args_to_strings(arguments)
+    def create_transaction_for_execute(self,
+                                       sender: IAddress,
+                                       contract: IAddress,
+                                       function: str,
+                                       gas_limit: int,
+                                       arguments: List[Any] = [],
+                                       native_transfer_amount: int = 0,
+                                       token_transfers: List[TokenTransfer] = []) -> Transaction:
+        number_of_tokens = len(token_transfers)
+
+        if native_transfer_amount and number_of_tokens:
+            raise BadUsageError("Can't send both native token and ESDT/NFT tokens")
+
+        token_computer = TokenComputer()
+        transfer_args: List[str] = []
+
+        if len(token_transfers) == 1:
+            transfer = token_transfers[0]
+
+            if token_computer.is_fungible(transfer.token):
+                transfer_args = self._build_args_for_esdt_transfer(transfer)
+                transfer_args.append(arg_to_string(function))
+                transfer_args.extend(args_to_strings(arguments))
+            else:
+                transfer_args = self._build_args_for_single_esdt_nft_transfer(transfer)
+                transfer_args.append(contract.hex())
+                contract = sender
```

Comment:
We should not have this line in the next PR, since it is not correct (even though, as part of a workaround, leads to the correct transaction). 

The receiver is set to the same address of the sender, yes, but it's not correct to say that "the contract address is the same as the sender".

Reply:
Modified it.